core/src/main/java/net/opentsdb/query/processor/groupby/BigSmallLinkedBlockingQueue.java

```diff
@@ -0,0 +1,91 @@
+package net.opentsdb.query.processor.groupby;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+public class BigSmallLinkedBlockingQueue<T> {
+
+  private final ConcurrentLinkedQueue<T> bigQ;
+  private final ConcurrentLinkedQueue<T> smallQ;
+
+  private volatile boolean takeFromBigQ;
+
+  private Predicate<T> bigPredicate;
+
+  private final AtomicInteger count = new AtomicInteger();
+
+  /** Lock held by take */
+  private final ReentrantLock readLock = new ReentrantLock();
+
+  private final Condition readSignal = readLock.newCondition();
+
+  public BigSmallLinkedBlockingQueue(Predicate<T> bigPredicate) {
+    this.bigPredicate = bigPredicate;
+    this.bigQ = new ConcurrentLinkedQueue();
+    this.smallQ = new ConcurrentLinkedQueue();
+  }
+
+  public int size() {
+    return count.get();
+  }
+
+  public void put(T t) {
+    boolean isBig = bigPredicate.test(t);
+    ConcurrentLinkedQueue<T> q = isBig ? bigQ : smallQ;
+
+    q.offer(t);
+    int c = count.getAndIncrement();
+
+    if (c == 0) {
```

Comment:
`>`

Reply:
or just delete the check as we always want to notify a thread when someone writes.